File: C/permutationTest.c
#include "permutationTest.h"
#include "utils.h"
#include <math.h>
#include <stdlib.h>
#include <stdio.h>

int compare(const void * a, const void * b)
{
	float v1 = *((float*)a);
	float v2 = *((float*)b);

    if (v1 < v2) return -1;
	if (v1 == v2) return 0;
	return 1; 
}

float evaluateStatistic(const TestData* data, const Permutation* permutation)
{
	float v = data->statistic(data, permutation);
	return v;
}

float evaluateStatisticMean(const TestData* data, const Permutation* permutation)
{
	float v1 = 0;
	float v2 = 0;
	for (unsigned a = 0; a < data->cutPoint; a++)
		v1 += data->sample[permutation->indicies[a]];

	v1 /= data->cutPoint;

	for (unsigned a = data->cutPoint; a < data->sampleLenght; a++)
		v2 += data->sample[permutation->indicies[a]];

	v2 /= data->sampleLenght - data->cutPoint;


	v1 -= v2;

	if (v1 < 0)
		v1 *= -1;

	return v1;
}

float hotelintTest(const TestData* data, const Permutation* toCopy)
{
	Vector mean;
	Vector mean2;
	for (unsigned b = 0; b < VECTOR_SIZE; b++)
	{
		mean[b] = 0;
		for (unsigned a = 0; a < data->cutPoint; a++)
			mean[b] += data->vectorSample[toCopy->indicies[a]][b]; 
		mean[b] /= data->sampleLenght;

		mean2[b] = 0;
		for (unsigned a = data->cutPoint; a < data->sampleLenght; a++)
			mean2[b] += data->vectorSample[toCopy->indicies[a]][b]; 
		mean2[b] /= data->sampleLenght;
	}

	float** matrix = (float**) malloc(VECTOR_SIZE * sizeof(float*));
	float** matrix2 = (float**) malloc(VECTOR_SIZE * sizeof(float*));

	for (unsigned a = 0; a < VECTOR_SIZE; a++)
	{
		matrix[a] = (float*) malloc(VECTOR_SIZE* sizeof(float));
		matrix2[a] = (float*) malloc(VECTOR_SIZE* sizeof(float));
	}
	for (unsigned a = 0; a < VECTOR_SIZE; a++)
		for (unsigned b = 0; b < VECTOR_SIZE; b++)
		{
			matrix[a][b] = 0;
			matrix2[a][b] = 0;
		}

	for (unsigned a = 0; a < data->cutPoint; a++)
	{
		for (unsigned b = 0; b < VECTOR_SIZE; b++)
		{
			for (unsigned c = 0; c < VECTOR_SIZE; c++)
			{
				float t = (data->vectorSample[toCopy->indicies[a]][b] - mean[b]);
				float r = (data->vectorSample[toCopy->indicies[a]][c] - mean[c]);
				matrix[b][c] += t * r; 
			}
		}
	}

	for (unsigned a = data->cutPoint; a < data->sampleLenght; a++)
	{
		for (unsigned b = 0; b < VECTOR_SIZE; b++)
		{
			for (unsigned c = 0; c < VECTOR_SIZE; c++)
			{
				float t = (data->vectorSample[toCopy->indicies[a]][b] - mean2[b]);
				float r = (data->vectorSample[toCopy->indicies[a]][c] - mean2[c]);
				matrix2[b][c] += t * r; 
			}
		}
	}

	for (unsigned a = 0; a < VECTOR_SIZE; a++)
		for (unsigned b = 0; b < VECTOR_SIZE; b++)
			matrix[a][b] = (matrix[a][b] + matrix2[a][b]) / (data->sampleLenght + 2);

	float determinante = Determinant(matrix, VECTOR_SIZE);


	for (unsigned a = 0; a < VECTOR_SIZE; a++)
		for (unsigned b = 0; b < VECTOR_SIZE; b++)
			matrix[a][b] /= determinante;

	Vector partial;

	for (unsigned a = 0; a < VECTOR_SIZE; a++)
	{
		partial[a] = 0;
		for (unsigned b = 0; b < VECTOR_SIZE; b++)
			partial[a] += matrix[a][b] * (mean[b] - mean2[b]);
	}

	float out = 0;
	for (unsigned a = 0; a < VECTOR_SIZE; a++)
		out += (mean[a] - mean2[a]) * partial[a];

	for (int a = 0; a < VECTOR_SIZE; a++)
	{
		free(matrix[a]);
		free(matrix2[a]);
	}
	free (matrix);
	free (matrix2);


	return out;

}

void merge(float** ranksOut, int left, int center, int right)
{
	int i = left;
	int j = center;
	int k = 0;
	float* b[right-left+1];

	while (i < center && j < right)
	{
		//printf ("%d %d\n", );
		if (*(ranksOut[i]) <= *(ranksOut[j]))
		{
			b[k] = ranksOut[i];	
			i++;
		}
		else
		{
			b[k] = ranksOut[j];
			j++;
		}
		k++;
	}

	while (i < center)
	{
		b[k] = ranksOut[i];
		i++;
		k++;
	}

	while (j < right)
	{
		b[k] = ranksOut[j];
		j++;
		k++;
	}

	for (k = left; k < right; k++)
		ranksOut[k] = b[k-left];
}

void calculateRanks(float** ranksOut, int left, int right)
{
	int center;
	if (left + 1>= right)
		return;

	center = (left+right)/2;
	calculateRanks(ranksOut, left, center);
	calculateRanks(ranksOut, center, right);

	merge(ranksOut, left, center, right);
}

float mannWitheyTest(const TestData* data, const Permutation* permutation)
{
	float** ranks = (float**)malloc(data->sampleLenght * sizeof(float*));	
	float* values = (float*)malloc(data->sampleLenght * sizeof(float));	
	float l1 = data->sampleLenght - data->cutPoint;
	float l0 = data->cutPoint;
	
	
	for (unsigned a = 0; a < data->sampleLenght; a++)
	{
		values[a] = data->sample[permutation->indicies[a]];
		ranks[a] = values + a;
	}
	calculateRanks(ranks, 0, data->sampleLenght);

	float u1 = 0;
	float u2 = 0;

	for (unsigned a = 0; a < data->sampleLenght; a++)
	{
		//printf("%d %f\n", ranks[a]-data->sample, *ranks[a]);
		if (ranks[a] - values < l0)
			u1 += a;
		else
			u2 += a;
	}

	u1 -= (l0*(l0+1))/2;

	//u2 -= ((data->sampleLenght-data->cutPoint)*((data->sampleLenght-data->cutPoint)+1))/2;
	//printf("%f %f\n", u1, u2);

	free(ranks);
	free(values);
	//if (u2 > u1)
	//	return u1;
	//else
	//	return u2;
	
	float mu = l0 * l1;
	float sd = sqrt(l0 * l1 * (l1+l0+1) / 12.0f); 
	u1 = fabs((u1-mu)/sd);
	//printf("%f\n", u1);
	return u1;
}

int permutationTestGetK(const TestData* data)
{
	return data->iterationsCount - ((int) (data->alpha * data->iterationsCount));
}

int runPermutationTest(const TestData* data)
{
    float statistics[data->iterationsCount];
	Permutation permutation = permutationCreate(data->sampleLenght);
	float sampleStatistic = evaluateStatistic(data, &permutation); //non permutated statistic
	
	unsigned equalOccurences = 0;
	unsigned higherOccurences = 0;
	int toBeReturned = 0;


	//calculate permutated statistics
	for (unsigned a = 0; a < data->iterationsCount; a++)
	{
		permutationRandomize(permutation);
		statistics[a] = evaluateStatistic(data, &permutation);

		if (statistics[a] == sampleStatistic)
			equalOccurences++;

		if (statistics[a] > sampleStatistic)
			higherOccurences++;
	}
	
	//sort statistics
    qsort(statistics, data->iterationsCount, sizeof (float), compare);
	int k = permutationTestGetK(data);
	
	if (statistics[k] > sampleStatistic)
		toBeReturned = 1;

	if (statistics[k] < sampleStatistic)
		toBeReturned = 0;

	if (statistics[k] == sampleStatistic)
	{
		float treshold = (data->alpha * data->iterationsCount) - higherOccurences;
		treshold /= equalOccurences;

		if (randomFloat() > treshold)
			toBeReturned = 0;
		else
			toBeReturned = 1;
	}

	permutationDestory(permutation);
	return toBeReturned;
}

void testDataClone(TestData* newOne, const TestData* toCopy)
{
	newOne->alpha = toCopy->alpha;
	newOne->sampleLenght = toCopy->sampleLenght;
	newOne->cutPoint = toCopy->cutPoint;
	newOne->iterationsCount = toCopy->iterationsCount;
	newOne->statistic = toCopy->statistic;
	if (toCopy->sample)
	{
		newOne->sample = (float*) malloc(sizeof(float) * toCopy->sampleLenght);
		for (unsigned a = 0; a < toCopy->sampleLenght; a++)
			newOne->sample[a] = toCopy->sample[a];
	}
	else
		newOne->sample = NULL;

	if (toCopy->vectorSample)
	{
		newOne->vectorSample = (Vector*) malloc(sizeof(Vector) *toCopy->sampleLenght);
		for (unsigned a = 0; a < toCopy->sampleLenght; a++)
			for (unsigned b = 0; b < VECTOR_SIZE; b++)
			newOne->vectorSample[a][b] = toCopy->vectorSample[a][b];
	}
	else
		newOne->vectorSample = NULL;

}

void destroyTestData(TestData* data)
{
	if (data->sample)
		free(data->sample);
	if (data->vectorSample)
		free(data->vectorSample);
}

File: matlab/permutationTest.m
function successPercent = permutationTest()
    alpha = 0.05;
    numberOfPermutation = 1000; %M
    sampleSize = 100; %N
    cutPoint = 50; %m 
    numberOfIterations = 1000;
    
    successPercent = 0;
    successData = zeros(1, numberOfPermutation - 50);
    builtInTestSuccess = 0;
    
    
    for a = 50:numberOfPermutation
        successPercent = 0;
        for t = 1:numberOfIterations
            [sample, sample2] = generateSample(sampleSize);       
            successPercent = successPercent + runPermutationTest(sample, sample2, alpha, a);
            %builtInTestSuccess = builtInTestSuccess + mwtest(sample, cutPoint, sampleSize);
            %builtInTestSuccess = builtInTestSuccess + ttest(sample, cutPoint, sampleSize);

        end
        successData(a - 49) = successPercent/numberOfIterations;
        disp(a);
    end
     
    y = linspace(50, numberOfPermutation, numberOfPermutation - 49);
    disp(size(y));
    disp(successData);
    disp(size(successData));
    %disp(builtInTestSuccess/numberOfIterations);
    scatter(y, successData);
    successPercent = successPercent/numberOfIterations*100;
end

function [sample, sample2] = generateSample(sampleSize)
    sample = normrnd(0, 1, [sampleSize/2, 1]);
    sample2 = normrnd(0, 1, [sampleSize/2, 1]);
end

function success = ttest(sample, sample2)
    success = 1-ttest2(sample, sample2);   
end

function success = mwtest(sample, sample2)
    [p, h] = ranksum(sample, sample2);
    success = 1-h;
end

function statistic = evaluateStatistic(sample, sample2)
    
    a = 0;
    b = 0;
    
    %for t = 1:cutPoint
    %    a = a + sample(t);
    %end
    
    
    %a = a / cutPoint;
    
    a = mean(sample);
    
    %for t = cutPoint+1:lenght
    %    b = b + sample(t);
    %end
   

    %b = b / (lenght - cutPoint);
    
    b = mean(sample);
    
    statistic = a - b;
    if (statistic < 0)
        statistic = statistic * -1;
    end
    
end

function statistic = evaluateStatisticMW(sample, sample2)
   
    [p, h, stat] = ranksum(sample, sample2);
    statistic = stat.zval;
    %disp(statistic);
    
    if (statistic < 0)
        statistic = statistic * -1;
    end
end

function h0Accepted = runPermutationTest(sample, sample2, alpha, numberOfP)

    statistics = zeros(1, numberOfP);    
    sampleStatistic = evaluateStatistic(sample, sample2);
    equalOccurences = 0;
    higherOccurences = 0;
    lenght = size(sample);
    lenght = lenght(1);
    lenght2 = size(sample2);
    lenght2 = lenght2(1);
    %disp(alpha);
    %disp(cutPoint);
    %disp(numberOfP);
    %disp(lenght);
    %disp('-');
    
    
    for t = 1:numberOfP
        statistics(t) = evaluateStatistic(sample(randperm(lenght)), sample2(randperm(lenght2))); 
        
        if (statistics(t) == sampleStatistic)
            equalOccurences = equalOccurences + 1;
        end
        
        if (statistics(t) > sampleStatistic)
            higherOccurences = higherOccurences + 1;
        end
    end
    
    statistics = sort(statistics);
    %disp(statistics);
    
    k = numberOfP - floor(alpha*numberOfP);
    %disp(statistics(k));
    %disp(sampleStatistic);
    if (statistics(k) > sampleStatistic)
        h0Accepted = 1;
       % disp(sampleStatistic);
       % disp(statistics(k));
        return
    end
    
    if (statistics(k) < sampleStatistic)
        h0Accepted = 0;
        return
    end
    
    treshold = (alpha*numberOfP) - higherOccurences;
    treshold = treshold / equalOccurences;
    
    if (rand() > treshold)
        h0Accepted = 0;
        return;
    end
        
    h0Accepted = 1;
    
end